src/plugins/setup.ts

```diff
@@ -0,0 +1,125 @@
+        const prevConfig = server.pluginConfigs.get(id)
+        const prevPlugin = prevConfig ? server.plugins.get(pluginConfig.plugin_id) : null
+
+        if (
+            prevConfig &&
+            pluginConfig.updated_at === prevConfig.updated_at &&
+            plugin?.updated_at == prevPlugin?.updated_at
+        ) {
+            pluginConfig.vm = prevConfig.vm
+        } else {
+            pluginConfig.vm = createLazyPluginVM()
+            pluginVMLoadPromises.push(loadPlugin(server, pluginConfig))
+        }
+    }
+
+    await Promise.all(pluginVMLoadPromises)
+
+    server.plugins = plugins
+    server.pluginConfigs = pluginConfigs
+    server.pluginConfigsPerTeam = pluginConfigsPerTeam
+
+    for (const teamId of server.pluginConfigsPerTeam.keys()) {
+        server.pluginConfigsPerTeam.get(teamId)?.sort((a, b) => a.order - b.order)
+    }
+
+    void loadSchedule(server)
+}
+
+async function loadPluginsFromDB(
+    server: PluginsServer
+): Promise<Pick<PluginsServer, 'plugins' | 'pluginConfigs' | 'pluginConfigsPerTeam'>> {
+    const pluginRows = await getPluginRows(server)
+    const plugins = new Map<PluginId, Plugin>()
+
+    for (const row of pluginRows) {
+        plugins.set(row.id, row)
+    }
+
+    const pluginAttachmentRows = await getPluginAttachmentRows(server)
+    const attachmentsPerConfig = new Map<TeamId, Record<string, PluginAttachment>>()
+    for (const row of pluginAttachmentRows) {
+        let attachments = attachmentsPerConfig.get(row.plugin_config_id)
+        if (!attachments) {
+            attachments = {}
+            attachmentsPerConfig.set(row.plugin_config_id, attachments)
+        }
+        attachments[row.key] = {
+            content_type: row.content_type,
+            file_name: row.file_name,
+            contents: row.contents,
+        }
+    }
+
+    const pluginConfigRows = await getPluginConfigRows(server)
+
+    const pluginConfigs = new Map<PluginConfigId, PluginConfig>()
+    const pluginConfigsPerTeam = new Map<TeamId, PluginConfig[]>()
+
+    for (const row of pluginConfigRows) {
+        const plugin = plugins.get(row.plugin_id)
+        if (!plugin) {
+            continue
+        }
+        const pluginConfig: PluginConfig = {
+            ...row,
+            plugin: plugin,
+            attachments: attachmentsPerConfig.get(row.id) || {},
+            vm: null,
+        }
+        pluginConfigs.set(row.id, pluginConfig)
+
+        if (!row.team_id) {
+            console.error(`🔴 PluginConfig(id=${row.id}) without team_id!`)
+            continue
+        }
+
+        let teamConfigs = pluginConfigsPerTeam.get(row.team_id)
+        if (!teamConfigs) {
+            teamConfigs = []
+            pluginConfigsPerTeam.set(row.team_id, teamConfigs)
+        }
+        teamConfigs.push(pluginConfig)
+    }
+
+    return { plugins, pluginConfigs, pluginConfigsPerTeam }
+}
+
+export async function loadSchedule(server: PluginsServer): Promise<void> {
```

Comment:
Not really since all callsites (e.g. worker) would then need to do `server.pluginSchedule = foobar` - this is a key piece in how the reloading logic works and I'd rather not leak this outside of this file.